File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2017/Y17D03.kt
package org.gristle.pdxpuzzles.advent.y2017

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.objects.Coord
import org.gristle.pdxpuzzles.utilities.enums.Nsew
import org.gristle.pdxpuzzles.utilities.math.isEven
import kotlin.math.ceil
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

class Y17D03(private val input: String) : Day {

    data class Turtle(var dir: Nsew, var pos: Coord, var vel: Int)

    override fun part1(): Int {
        val squareRoot = ceil(sqrt(input.toFloat())).toInt().let { if (it.isEven()) it + 1 else it }
        val furthest = (squareRoot / 2) * 2
        val br = squareRoot * squareRoot
        val diff = (br - input.toInt()) % furthest
        return furthest - min(diff, furthest / 2) + max(0, diff - furthest / 2)
    }

    override fun part2(): Int {
        val t = Turtle(Nsew.SOUTH, Coord.ORIGIN, 0)
        val space = mutableMapOf(t.pos to 1)
        while (true) {
            t.dir = t.dir.left()
            if (t.dir == Nsew.EAST || t.dir == Nsew.WEST) t.vel++
            for (i in 1..t.vel) {
                t.pos = t.pos.move(t.dir)
                val squareVal = t.pos.getNeighbors(true).sumOf { space[it] ?: 0 }
                if (squareVal > input.toInt()) return squareVal
                space[t.pos] = squareVal
            }
        }
    }
}

fun main() = Day.runDay(Y17D03::class)

//    Class creation: 9ms
//    Part 1: 552 (0ms)
//    Part 2: 330785 (3ms)
//    Total time: 13ms

File: rust/advent/src/bin/2017/y17d03.rs
use std::cmp::{max, min};

use advent::utilities::get_input::get_input;
use rustc_hash::FxHashMap;
use utilities::{enums::cardinals::Cardinal, structs::{coord::Coord2, stopwatch::{ReportDuration, Stopwatch}}};

type Input = usize;
type Output = usize;
type Pos = Coord2;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(17, 3).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(input), stopwatch.lap().report());
    println!("2. {} ({})", part2(input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn parse_input(input: &str) -> Input {
    input.parse().unwrap()
}

fn part1(square: Input) -> Output {
    let square_root = (square as f64).sqrt().ceil() as usize;
    let square_root = square_root + if square_root & 1 == 0 { 1 } else { 0 };
    let farthest = square_root / 2 * 2;
    let br = square_root * square_root;
    let diff = (br - square) % farthest;
    farthest - min(diff, farthest / 2) + max(0, diff - farthest / 2)
}

struct Turtle {
    dir: Cardinal,
    pos: Pos,
    vel: usize,
}
fn part2(input: Input) -> Output {
    let mut turtle = Turtle { dir: Cardinal::South, pos: Pos::origin(), vel: 0 };
    let mut space: FxHashMap<Pos, usize> = FxHashMap::default();
    space.insert(Pos::origin(), 1);
    loop {
        turtle.dir = turtle.dir.left();
        if turtle.dir == Cardinal::East || turtle.dir == Cardinal::West {
            turtle.vel += 1;
        }
        for _ in 1..=turtle.vel {
            turtle.pos = turtle.pos.move_direction(turtle.dir, 1).unwrap();
            let square_val: usize = turtle.pos.adjacent(true).into_iter()
                .map(|adj| space.get(&adj).copied().unwrap_or_default())
                .sum();
            if square_val > input {
                return square_val;
            }
            space.insert(turtle.pos, square_val);
        }
    }
}


#[test]
fn default() {
    let input = get_input(17, 3).unwrap();
    let input = parse_input(&input);
    assert_eq!(552, part1(input));
    assert_eq!(330785, part2(input));
}

// Input parsed (12μs)
// 1. 552 (5μs)
// 2. 330785 (11μs)
// Total: 30μs
